Source: https://github.com/ony/ddclient
Upstream: https://sourceforge.net/p/ddclient/bugs/74/ https://github.com/wimpunk/ddclient/pull/28
Reason: Upstream dead for 2 years already. Fork wimpunk accepts pull requests very slowly.

From 11e6304671ddc7c45a14a6a6f6278ab302f2e807 Mon Sep 17 00:00:00 2001
From: Mykola Orliuk <virkony@gmail.com>
Date: Fri, 23 Oct 2015 12:16:10 +0300
Subject: [PATCH] freedns: send IP address

http://freedns.afraid.org/ automatically detects external IP. To
override that IP it is possible to supply address=IP as additional
parameter.
---
 ddclient | 22 +++++++++++++++++++++-
 1 file changed, 21 insertions(+), 1 deletion(-)

diff --git a/ddclient b/ddclient
index 7abb5dd..ae94972 100755
--- a/ddclient
+++ b/ddclient
@@ -1920,6 +1920,26 @@ EOM
 	import JSON::Any;
 }
 ######################################################################
+## urladd
+######################################################################
+sub urladd {
+    my $url      = shift || '';
+    debug("url    = %s", $url);
+    unless ($url =~ /(\?[^#]*)?(#.*)?$/) {
+	    failed("invalid url: %s", $url);
+	    return undef;
+	}
+	return $url if (@_ == 0);
+	my ($base, $query, $fragment) = ($`, $1, $2);
+	if (!$query) {
+		$query = '?' . shift;
+	}
+	foreach my $param (@_) {
+		$query .= '&' . $param;
+	}
+	return $base . $query . $fragment;
+}
+######################################################################
 ## geturl
 ######################################################################
 sub geturl {
@@ -3739,7 +3759,7 @@ sub nic_freedns_update {
 	    $config{$h}{'status'} = 'good'; 
 	    success("update not necessary %s: good: IP address already set to %s", $h, $ip); 
 	} else {
-	    my $reply = geturl(opt('proxy'), $freedns_hosts{$h}->[2]);
+	    my $reply = geturl(opt('proxy'), urladd($freedns_hosts{$h}->[2], "address=$ip"));
 	    if (!defined($reply) || !$reply) {
 	        failed("updating %s: Could not connect to %s.", $h, $freedns_hosts{$h}->[2]);
 		last;
-- 
2.6.3

